tree/dataframe/inc/ROOT/RDF/ActionHelpers.hxx

```diff
@@ -1800,23 +1822,169 @@ public:
+      : fFileName(filename),
+        fNTupleName(ntuplename),
+        fOptions(options),
+        fLoopManager(lm),
+        fInputFieldNames(vfnames),
+        fOutputFieldNames(ReplaceDotWithUnderscore(fnames)),
+        fIsDefine(std::move(isDefine))
+   {
+      ValidateSnapshotRNTupleOutput(fOptions, fNTupleName, fFileName);
+   }
+
+   SnapshotRNTupleHelper(const SnapshotRNTupleHelper &) = delete;
+   SnapshotRNTupleHelper(SnapshotRNTupleHelper &&) = default;
+   ~SnapshotRNTupleHelper()
+   {
+      if (!fNTupleName.empty() && !fLoopManager->GetDataSource() && fOptions.fLazy)
+         Warning("Snapshot", "A lazy Snapshot action was booked but never triggered.");
+   }
+
+   void InitTask(TTreeReader *, unsigned int /* slot */) {}
+
+   void Exec(unsigned int /* slot */, ColTypes &...values)
+   {
+      using ind_t = std::index_sequence_for<ColTypes...>;
+
+      SetFields(values..., ind_t{});
+      fWriter->Fill();
+   }
+
+   template <std::size_t... S>
+   void SetFields(ColTypes &...values, std::index_sequence<S...> /*dummy*/)
+   {
+      int expander[] = {(SetFieldsHelper(values, fOutputFieldNames[S], fOutputEntry), 0)..., 0};
+      (void)expander; // avoid unused variable warnings for older compilers (gcc 14.1)
+   }
+
+   void Initialize()
+   {
+      using ind_t = std::index_sequence_for<ColTypes...>;
+
+      auto model = ROOT::Experimental::RNTupleModel::Create();
+      MakeFields(*model, ind_t{});
+      fOutputEntry = &model->GetDefaultEntry();
+
+      fOutFile = std::unique_ptr<TFile>(
+         TFile::Open(fFileName.c_str(), fOptions.fMode.c_str(), /*ftitle=*/fFileName.c_str() /* , cs */));
+      if (!fOutFile)
+         throw std::runtime_error("Snapshot: could not create output file " + fFileName);
+
+      ROOT::Experimental::RNTupleWriteOptions writeOptions;
+      writeOptions.SetCompression(fOptions.fCompressionAlgorithm, fOptions.fCompressionLevel);
+
+      TString checkupdate = fOptions.fMode;
+      checkupdate.ToLower();
+
+      if (checkupdate == "update") {
+         fWriter = ROOT::Experimental::RNTupleWriter::Append(std::move(model), fNTupleName, *fOutFile, writeOptions);
+      } else {
+         // If the output file is not closed before the RNTupleWriter is created, no actual data will be written and
+         // the file will be empty.
+         fOutFile->Close();
+         fWriter = ROOT::Experimental::RNTupleWriter::Recreate(std::move(model), fNTupleName, fFileName, writeOptions);
+      }
+   }
+
+   template <std::size_t... S>
+   void MakeFields(ROOT::Experimental::RNTupleModel &model, std::index_sequence<S...> /*dummy*/)
+   {
+      int expander[] = {(model.MakeField<ColTypes>(fOutputFieldNames[S]), 0)..., 0};
+      (void)expander; // avoid unused variable warnings for older compilers (gcc 14.1)
+   }
+
+   void Finalize()
+   {
+      {
+         fWriter.reset();
+      }
+      fLoopManager->SetDataSource(std::make_unique<ROOT::Experimental::RNTupleDS>(fNTupleName, fFileName));
```

Comment:
I know it's been a while, but why are we replacing the data source? I don't see this done for the existing `SnapshotTTreeHelper`, but maybe it is hidden somewhere because of how TTree is special in RDF...

I'm asking because I guess this one / the reason for the compatibility matrix in the PR summary. Also I'm not sure if this change of the data source is actually sound, I think with some `Redefine`s I can construct cases where filters are not idempotent and thus a snapshot + replacing the data source might change results.

Reply:
To answer myself: `Snapshot` directly calls `CreateLMFromTTree` with a non-existing file name that is then later written to during the action. See also Vincenzo's comment https://github.com/root-project/root/pull/15750/files#r1741566386